lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,145 @@
+	fileSharedWriter *FileSharedWriter
+
+	// initLock protects global init of the file shared writer.
+	initLock sync.Mutex
+)
+
+// FileSharedWriter is similar to SharedWriter except that it requires a `os.File` instead of a `io.Writer`.
+// This is to allow the File reopen required by logrotate and similar tools.
+type FileSharedWriter struct {
+	logFile  atomic.Pointer[os.File]
+	fileFlag int
+	fileMode fs.FileMode
+
+	lock sync.Mutex
+}
+
+// InitFileSharedWriter wraps the provided [os.File] in a writer that is thread safe,
+// with ability to enable filesystem notification watch and reopen file on specific events.
+func InitFileSharedWriter(logFile *os.File, flag int, mode fs.FileMode) (io.Writer, error) {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if fileSharedWriter != nil {
+		return nil, trace.BadParameter("file shared writer already initialized")
+	}
+	fileSharedWriter = &FileSharedWriter{fileFlag: flag, fileMode: mode}
+	fileSharedWriter.logFile.Store(logFile)
+
+	return fileSharedWriter, nil
+}
+
+// GetFileSharedWriter returns instance of the file shared writer.
+func GetFileSharedWriter() *FileSharedWriter {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	return fileSharedWriter
+}
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.logFile.Load().Write(b)
+}
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	logFile, err := os.OpenFile(s.logFile.Load().Name(), s.fileFlag, s.fileMode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	prevLogFile := s.logFile.Swap(logFile)
+	if err := prevLogFile.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// RunWatcherFunc spawns goroutine with the watcher loop to consume events of renaming
+// or removing the log file to trigger the action function when event appeared.
+func (s *FileSharedWriter) RunWatcherFunc(ctx context.Context, action func() error) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if s.logFile.Load().Name() == event.Name && (event.Has(fsnotify.Rename) || event.Has(fsnotify.Remove)) {
+					slog.DebugContext(ctx, "Log file was moved/removed, reopen new one", "file", event.Name)
+					if err := action(); err != nil {
+						slog.ErrorContext(ctx, "Failed to reopen new file", "error", err, "file", event.Name)
+						continue
+					}
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				slog.ErrorContext(ctx, "Error received on logger watcher", "error", err)
+			case <-ctx.Done():
+				if err := watcher.Close(); err != nil {
+					slog.ErrorContext(context.Background(), "Failed to close file watcher", "error", err)
+				}
```

Comment:
That's a busyloop until one of events and errors is closed; ability to ever stop watching the file notwithstanding, why not tie the context to the watcher close with a `context.AfterFunc`?